[CostModel] Remove optional from InstructionCost::getValue()

davemgreen opened the pull request:
InstructionCost is already an optional value, containing an Invalid state that can be checked with isValid(). There is little point in returning another optional from getValue(). Most uses do not make use of it being a std::optional, dereferencing the value directly (either isValid has been checked previously or the Cost is assumed to be valid). The one case that does in AMDGPU used value_or which has been replaced by a new getValueOr(CostType) function.

alexey-bataev commented on llvm/lib/Target/AMDGPU/AMDGPUSplitModule.cpp:
```diff
@@ -205,8 +205,7 @@ static CostType calculateFunctionCosts(GetTTIFn GetTTI, Module &M,
         // Assume expensive if we can't tell the cost of an instruction.
-        CostType CostVal =
-            Cost.getValue().value_or(TargetTransformInfo::TCC_Expensive);
+        CostType CostVal = Cost.getValueOr(TargetTransformInfo::TCC_Expensive);
```
Maybe just add a check here and remove `getValueOr()`?

Pierre-vh replied:
> Maybe just add a check here and remove getValueOr()?

I wrote this assuming that Cost may not be possible to calculate for every instruction (maybe target intrinsics?).
If the cost is always calculated and cannot really fail then sure, we can replace this with an assert

davemgreen replied:
I have removed getValueOr and performed the check inline.

(It is a good thing to check for invalid costs even if they are unlikely to come up, to make things more robust. They shouldn't come up in already-present llvm-ir though ideally, most IR get some valid cost and the invalid costs are used for things that cannot be selected/legalized).

RKSimon replied:
(side note) I'd dearly like to get rid of TCC_Expensive entirely - the value of 4 doesn't really mean that much, especially for different CostKinds, and saying that it represents the cost of a typical x86 div instruction is just wrong.

davemgreen replied:
That sounds sensible to me, I might personally remove all the TCC_ costs as "0" or "1" are fairly self-explanatory numbers. TCC_Expensive would need to be replaced by something. I recently went and measured the cost of a divide, using dynamorio to collect the distributions of numerators/denominators and checking the latency of divides with those values. The real value would vary depending on how you weighted it (there were a lot of 1/1 divides). Having it as 4 at least makes a starting-point, and targets can override it when they come up with a better value.

davemgreen committed:
[CostModel] Remove optional from InstructionCost::getValue()

InstructionCost is already an optional value, containing an Invalid state that
can be checked with isValid(). There is little point in returning another
optional from getValue(). Most uses do not make use of it being a
std::optional, dereferencing the value directly (either isValid has been
checked previously or the Cost is assumed to be valid). The one case that does
in AMDGPU used value_or which has been replaced by a new getValueOr(CostType)
function.
